File: app/src/main/java/pl/mosenko/sunnypodlaskie/mvp/weatherdatalist/WeatherDataListFragment.java
package pl.mosenko.sunnypodlaskie.mvp.weatherdatalist;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.ViewGroup;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import com.hannesdorfmann.mosby3.mvp.MvpFragment;

import java.util.List;

import butterknife.BindView;
import butterknife.BindViews;
import butterknife.ButterKnife;
import butterknife.Unbinder;
import pl.aprilapps.switcher.Switcher;
import pl.mosenko.sunnypodlaskie.BuildConfig;
import pl.mosenko.sunnypodlaskie.R;
import pl.mosenko.sunnypodlaskie.persistence.entities.WeatherDataEntity;

/**
 * Created by syk on 20.05.17.
 */

public class WeatherDataListFragment extends MvpFragment<WeatherDataListContract.View, WeatherDataListContract.Presenter>
        implements WeatherDataListContract.View, SwipeRefreshLayout.OnRefreshListener, WeatherDataListAdaper.WeatherDataClickedListener {

    private static final String TAG = WeatherDataListFragment.class.getSimpleName();

    @BindView(R.id.WeatherDataListFragment_TextView_Empty)
    TextView textViewEmpty;
    @BindView(R.id.WeatherDataListFragment_ProgressBar)
    ProgressBar progressBarLoading;
    @BindView(R.id.WeatherDataListFragment_RecyclerView)
    RecyclerView recyclerView;
    @BindView(R.id.FragmentWeatherDataList_TextView_Error)
    TextView textViewError;

    @BindView(R.id.WeatherDataListFragment_SwipeRefreshLayout_Empty)
    SwipeRefreshLayout swipeRefreshLayoutEmpty;
    @BindView(R.id.WeatherDataListFragment_SwipeRefreshLayout_Error)
    SwipeRefreshLayout swipeRefreshLayoutError;
    @BindView(R.id.FragmentWeatherDataList_SwipeRefreshLayout_Content)
    SwipeRefreshLayout swipeRefreshLayoutContent;
    @BindViews({
            R.id.WeatherDataListFragment_SwipeRefreshLayout_Empty,
            R.id.WeatherDataListFragment_SwipeRefreshLayout_Error,
            R.id.FragmentWeatherDataList_SwipeRefreshLayout_Content
    })
    List<SwipeRefreshLayout> swipeRefreshLayouts;

    static final ButterKnife.Setter<SwipeRefreshLayout, SwipeRefreshLayout.OnRefreshListener> SET_SWIPE_REFRESH_LAYOUT_LISTENER =
            (view, listener, index) -> view.setOnRefreshListener(listener);
    static final ButterKnife.Action<SwipeRefreshLayout>  STOP_REFRESHING = (view, index) -> view.setRefreshing(false);
    private static final Callback sDummyCallback = (id) -> {};
    private Switcher mSwitcher;
    private Callback mCallback = sDummyCallback;
    private WeatherDataListAdaper mWeatherDataListAdaper;
    private Unbinder mUnbinder;

    public WeatherDataListFragment() {
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        tryToInitializeCallbackField(getActivity());
    }

    private void tryToInitializeCallbackField(Context context) {
        try {
            mCallback = (Callback) context;
        } catch (ClassCastException e) {
            throw new ClassCastException(context.toString()
                    + " must implement Callback");
        }
    }

    @Nullable
    @Override
    public android.view.View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, Bundle savedInstanceState) {
        android.view.View rootView = inflateFragmentLayout(inflater, container);
        bindGraphicalComponents(rootView);
        return rootView;
    }

    @Override
    public void onViewCreated(android.view.View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        configureSwitcherView();
        customizeRecyclerView();
        customizeSwipeRefreshLayout();
    }

    private void configureSwitcherView() {
        mSwitcher = new Switcher.Builder(getContext())
                .addContentView(swipeRefreshLayoutContent)
                .addEmptyView(swipeRefreshLayoutEmpty)
                .addProgressView(progressBarLoading)
                .addErrorView(swipeRefreshLayoutError)
                .build();
    }

    private void customizeRecyclerView() {
        LinearLayoutManager layoutManager =
                new LinearLayoutManager(getActivity(), LinearLayoutManager.VERTICAL, false);
        recyclerView.setLayoutManager(layoutManager);
        recyclerView.setHasFixedSize(true);
        mWeatherDataListAdaper = new WeatherDataListAdaper(getActivity(), this);
        recyclerView.setAdapter(mWeatherDataListAdaper);
    }

    private void bindGraphicalComponents(android.view.View rootView) {
        mUnbinder = ButterKnife.bind(this, rootView);
    }

    private android.view.View inflateFragmentLayout(LayoutInflater inflater, @Nullable ViewGroup container) {
        return inflater.inflate(R.layout.fragment_weather_data_list, container, false);
    }

    @Override
    public void showEmpty() {
        ButterKnife.apply(swipeRefreshLayouts, STOP_REFRESHING);
        mSwitcher.showEmptyView();
    }

    @Override
    public void showLoading(boolean pullToRefresh) {
        if (!pullToRefresh) {
            mSwitcher.showProgressView();
        }
    }

    @Override
    public void showContent() {
        ButterKnife.apply(swipeRefreshLayouts, STOP_REFRESHING);
        mSwitcher.showContentView();
    }

    @Override
    public void showError(Throwable throwable) {
        showError(throwable, true);
    }

    @Override
    public void showDataWithoutInternetUpdatedMessage() {
        Toast.makeText(getActivity(), R.string.message_no_connection, Toast.LENGTH_SHORT).show();
    }

    @Override
    public void showError(Throwable e, boolean pullToRefresh) {
        if (BuildConfig.DEBUG) {
            Log.e(TAG, e.getMessage(), e);
        }
        if (pullToRefresh) {
            ButterKnife.apply(swipeRefreshLayouts, STOP_REFRESHING);
        }
        mSwitcher.showErrorView();
}

    @Override
    public void setData(List<WeatherDataEntity> weatherDataEntityList) {
        mWeatherDataListAdaper.swapWeatherList(weatherDataEntityList);
    }

    private void customizeSwipeRefreshLayout() {
        ButterKnife.apply(swipeRefreshLayouts, SET_SWIPE_REFRESH_LAYOUT_LISTENER, this);
        for (SwipeRefreshLayout swipeRefreshLayout : swipeRefreshLayouts) {
            swipeRefreshLayout.setColorSchemeResources(
                    R.color.colorAccent,
                    R.color.activated,
                    R.color.colorPrimary,
                    R.color.colorPrimaryDark);
        }
    }

    @Override
    public void onRefresh() {
        loadData(true);
    }

    @NonNull
    @Override
    public WeatherDataListContract.Presenter createPresenter() {
        return new WeatherDataListPresenterImpl();
    }

    @Override
    public void onWeatherDataItemClick(long id) {
            mCallback.onItemSelected(id);
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mCallback = sDummyCallback;
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        mUnbinder.unbind();
    }

    @Override
    public void onResume() {
        super.onResume();
        getPresenter().onResume();
    }

    @Override
    public void onPause() {
        super.onPause();
        getPresenter().onPause();
    }

    @Override
    public void loadData(boolean pullToRefresh) {
        getPresenter().loadData(pullToRefresh);
    }

    public interface Callback {
        void onItemSelected(long weatherDataId);
    }
}

File: app/src/main/java/pl/mosenko/sunnypodlaskie/mvp/weatherdatalist/WeatherDataListFragment.kt
package pl.mosenko.sunnypodlaskie.mvp.weatherdatalist

import android.os.Bundle
import android.util.Log
import android.view.*
import android.widget.Toast
import androidx.core.os.bundleOf
import androidx.fragment.app.Fragment
import androidx.navigation.Navigation
import androidx.navigation.fragment.NavHostFragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout.OnRefreshListener
import org.koin.android.ext.android.inject
import pl.aprilapps.switcher.Switcher
import pl.mosenko.sunnypodlaskie.BuildConfig
import pl.mosenko.sunnypodlaskie.R
import pl.mosenko.sunnypodlaskie.databinding.FragmentWeatherDataListBinding
import pl.mosenko.sunnypodlaskie.mvp.weatherdatadetail.WeatherDataDetailFragment
import pl.mosenko.sunnypodlaskie.mvp.weatherdatalist.WeatherDataListAdaper.WeatherDataClickedListener
import pl.mosenko.sunnypodlaskie.persistence.entities.WeatherDataEntity

/**
 * Created by syk on 20.05.17.
 */
class WeatherDataListFragment :
    Fragment(),
    WeatherDataListContract.View, OnRefreshListener, WeatherDataClickedListener {

    private val presenter: WeatherDataListContract.Presenter by inject()
    private val isTablet by lazy { resources.getBoolean(R.bool.is_tablet) }

    private var _binding: FragmentWeatherDataListBinding? = null
    private val binding get() = _binding!!

    private lateinit var swipeRefreshLayouts: List<SwipeRefreshLayout>
    private lateinit var switcher: Switcher
    private lateinit var weatherDataListAdaper: WeatherDataListAdaper

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentWeatherDataListBinding.inflate(inflater, container, false)
        swipeRefreshLayouts = listOf(
            binding.incContent.srlContent, binding.incEmpty.srlEmpty,
            binding.incError.srlError
        )
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setHasOptionsMenu(true)
        presenter.attachView(this)
        configureSwitcherView()
        customizeRecyclerView()
        customizeSwipeRefreshLayout()
    }

    private fun configureSwitcherView() {
        switcher = Switcher.Builder(requireContext())
            .addContentView(binding.incContent.srlContent)
            .addEmptyView(binding.incEmpty.srlEmpty)
            .addProgressView(binding.pbLoading)
            .addErrorView(binding.incError.srlError)
            .build()
    }

    private fun customizeRecyclerView() {
        val layoutManager = LinearLayoutManager(activity, LinearLayoutManager.VERTICAL, false)
        binding.incContent.rvContent.let {
            it.layoutManager = layoutManager
            it.setHasFixedSize(true)
            weatherDataListAdaper = WeatherDataListAdaper(this)
            it.adapter = weatherDataListAdaper
        }
    }

    override fun showEmpty() {
        stopRefreshLayouts()
        switcher.showEmptyView()
    }

    private fun stopRefreshLayouts() {
        swipeRefreshLayouts.forEach {
            it.isRefreshing = false
        }
    }

    override fun showLoading(pullToRefresh: Boolean) {
        if (!pullToRefresh) {
            switcher.showProgressView()
        }
    }

    override fun showContent() {
        stopRefreshLayouts()
        switcher.showContentView()
    }

    override fun showError(throwable: Throwable) {
        showError(throwable, true)
    }

    override fun showDataWithoutInternetUpdatedMessage() {
        Toast.makeText(activity, R.string.message_no_connection, Toast.LENGTH_SHORT).show()
    }

    override fun showError(e: Throwable, pullToRefresh: Boolean) {
        if (BuildConfig.DEBUG) {
            Log.e(TAG, e.message, e)
        }
        if (pullToRefresh) {
            stopRefreshLayouts()
        }
        switcher.showErrorView()
    }

    override fun setData(data: MutableList<WeatherDataEntity>) {
        weatherDataListAdaper.swapWeatherList(data)
    }

    private fun customizeSwipeRefreshLayout() {
        setSwipeRefreshLayoutListener()
        for (swipeRefreshLayout in swipeRefreshLayouts) {
            swipeRefreshLayout.setColorSchemeResources(
                R.color.colorAccent,
                R.color.activated,
                R.color.colorPrimary,
                R.color.colorPrimaryDark
            )
        }
    }

    private fun setSwipeRefreshLayoutListener() {
        swipeRefreshLayouts.forEach {
            it.setOnRefreshListener(this)
        }
    }

    override fun onRefresh() {
        loadData(true)
    }

    override fun onWeatherDataItemClick(weatherDataId: Long) {
        if (isTablet) {
            val navHostFragment = childFragmentManager.findFragmentById(
                R.id.fcv_container
            ) as NavHostFragment
            navHostFragment.navController.navigate(
                R.id.weatherDataDetailFragmentTablet,
                bundleOf(WeatherDataDetailFragment.ARG_WEATHER_DATA_ID to weatherDataId)
            )
        } else {
            Navigation.findNavController(binding.root).navigate(
                WeatherDataListFragmentDirections.navigateToWeatherDataDetailFragment(weatherDataId)
            )
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        presenter.detachView()
        _binding = null
    }

    override fun onResume() {
        super.onResume()
        presenter.onResume()
    }

    override fun onPause() {
        super.onPause()
        presenter.onPause()
    }

    override fun loadData(pullToRefresh: Boolean) {
        presenter.loadData(pullToRefresh)
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        menu.clear()
        inflater.inflate(R.menu.main_menu, menu)
        super.onCreateOptionsMenu(menu, inflater)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId
        if (id == R.id.action_settings) {
            Navigation.findNavController(binding.root).navigate(R.id.navigateToSettingsFragment)
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    companion object {
        private val TAG = WeatherDataListFragment::class.java.simpleName
    }
}
